### PR TITLE
NODE-2568 Transaction state snapshot storage

### PR DESCRIPTION
**Pull request by xrtm000**

### CODE REVIEW

File: node/src/main/scala/com/wavesplatform/database/RocksDBWriter.scala [hunk 1]
```diff
@@ -531,17 +520,23 @@ class RocksDBWriter(
...
-      for ((alias, addressId) <- aliases) {
-        rw.put(Keys.addressIdOfAlias(alias), Some(addressId))
+      for ((alias, address) <- snapshot.aliases) {
+        val key   = Keys.addressIdOfAlias(alias)
+        val value = addressIdWithFallback(address, newAddresses)
+        rw.put(key, Some(value))
+      }
+
+      for ((assetId, sponsorship) <- snapshot.sponsorships) {
+        rw.put(Keys.sponsorship(assetId)(height), SponsorshipValue(sponsorship.minFee))
```
Comment (ivan-mashonskiy) [inline node/src/main/scala/com/wavesplatform/database/RocksDBWriter.scala]:
```suggestion
        rw.put(Keys.sponsorship(assetId)(height), sponsorship)
```

File: node/src/test/scala/com/wavesplatform/test/DomainPresets.scala [hunk 2]
```diff
@@ -100,7 +100,7 @@ object DomainPresets {
       featuresSettings = RideV6.featuresSettings.copy(autoShutdownOnUnsupportedFeature = false)
     )
 
-  val TransactionStateSnapshot: WavesSettings = ConsensusImprovements.addFeatures(BlockchainFeatures.TransactionStateSnapshot)
+  val TransactionStateSnapshotP: WavesSettings = ConsensusImprovements.addFeatures(BlockchainFeatures.TransactionStateSnapshot)
```
Comment (ivan-mashonskiy) [inline node/src/test/scala/com/wavesplatform/test/DomainPresets.scala]: Is it typo or it's intended?

File: node/src/main/scala/com/wavesplatform/state/StateSnapshot.scala [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+    scriptResults: Map[ByteStr, InvokeScriptResult] = Map(),
+    ethereumTransactionMeta: Map[ByteStr, EthereumTransactionMeta] = Map(),
+    scriptsComplexity: Long = 0
+) {
+  import com.wavesplatform.protobuf.snapshot.TransactionStateSnapshot as S
+
+  def toProtobuf(txSucceeded: Boolean): TransactionStateSnapshot =
+    TransactionStateSnapshot(
+      balances.map { case ((address, asset), balance) =>
+        S.Balance(address.toByteString, Some(Amount(asset.fold(ByteString.EMPTY)(_.id.toByteString), balance)))
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/state/StateSnapshot.scala]: `PBAmounts.toPBAssetId`?

File: node/src/main/scala/com/wavesplatform/state/VolumeAndFee.scala [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+case class VolumeAndFee(volume: Long, fee: Long)
+
+object VolumeAndFee {
+  val empty: VolumeAndFee = VolumeAndFee(0, 0)
+
+  implicit val m: Monoid[VolumeAndFee] = new Monoid[VolumeAndFee] {
+    override def empty: VolumeAndFee = VolumeAndFee.empty
+
+    override def combine(x: VolumeAndFee, y: VolumeAndFee): VolumeAndFee =
+      VolumeAndFee(x.volume + y.volume, x.fee + y.fee)
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/state/VolumeAndFee.scala]: `safeSum`?

File: node-it/src/test/scala/com/wavesplatform/it/sync/smartcontract/EstimatorTestSuite.scala [hunk 5]
```diff
@@ -111,7 +112,9 @@ class EstimatorTestSuite extends BaseTransactionSuite with CancelAfterFailure {
   }
 
   test("can still invoke account and asset scripts after estimator v1 precheck activation") {
-    sender.invokeScript(callerAcc, smartAcc.toAddress.toString, waitForTx = true)._1.id
+    // after snapshot implementation only evaluated complexity is used for dApps
+    // so this call fails
+    // sender.invokeScript(callerAcc, smartAcc.toAddress.toString, waitForTx = true)._1.id
     sender.transfer(smartAcc, callerAcc.toAddress.toString, 1, minFee + 2 * smartFee, Some(issuedAssetId), None, waitForTx = true)
   }
```
Comment (vsuharnikov) [inline node-it/src/test/scala/com/wavesplatform/it/sync/smartcontract/EstimatorTestSuite.scala:129]: So these tests should be renamed? Also probably we should intercept a fail and match the exception, at least the fact of it?

File: node/src/main/scala/com/wavesplatform/database/Caches.scala [hunk 6]
```diff
@@ -205,32 +202,24 @@ abstract class Caches extends Blockchain with Storage {
...
+      accountScripts: Map[AddressId, Option[AccountScriptInfo]],
+      stateHash: StateHashBuilder.Result
   ): Unit
 
-  override def append(diff: Diff, carryFee: Long, totalFee: Long, reward: Option[Long], hitSource: ByteStr, block: Block): Unit = {
-    val newHeight = current.height + 1
-    val newScore  = block.blockScore() + current.score
+  override def append(snapshot: StateSnapshot, carryFee: Long, totalFee: Long, reward: Option[Long], hitSource: ByteStr, block: Block): Unit = {
+    val newHeight    = current.height + 1
+    val newAddresses = Set.newBuilder[Address]
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/database/Caches.scala]: We use this Set only in this function. `mutable.Set` looks more straightforward here.

File: node/src/test/scala/com/wavesplatform/state/diffs/smart/predef/BrokenUnicodeTest.scala [hunk 7]
```diff
@@ -200,13 +200,13 @@ class BrokenUnicodeTest extends PropSpec with WithDomain with EitherValues {
...
-    d.appendBlock(checkNoFix*)
-    checkNoFix.foreach(tx => d.blockchain.transactionSucceeded(tx.id.value()) shouldBe true)
-  }
+//  private def assertNoFix(d: Domain): Unit = {
+//    val (genesisTxs, setNoFix, _, checkNoFix, _) = scenario(V5)
+//    d.appendBlock(genesisTxs*)
+//    d.appendBlock(setNoFix*)
+//    d.appendBlock(checkNoFix*)
+//    checkNoFix.foreach(tx => d.blockchain.transactionSucceeded(tx.id.value()) shouldBe true)
+//  }
```
Comment (vsuharnikov) [inline node/src/test/scala/com/wavesplatform/state/diffs/smart/predef/BrokenUnicodeTest.scala]: Remove?

File: node/src/main/scala/com/wavesplatform/transaction/smart/WavesEnvironment.scala [hunk 8]
```diff
@@ -353,6 +354,7 @@ object DAppEnvironment {
   }
 }
 
+// todo move to separate class
```
Comment (vsuharnikov) [inline node/src/main/scala/com/wavesplatform/transaction/smart/WavesEnvironment.scala:358]: ?

File: node/src/test/scala/com/wavesplatform/state/diffs/BlockDifferDetailedDiffTest.scala [hunk 9]
```diff
@@ -1,131 +1,132 @@
...
-            assertDetailedDiff(history, block, ngFs) { case (_, DetailedDiff(parentDiff, _)) =>
-              parentDiff.portfolios(ngMiner).balance shouldBe transactionFee // 60% + 40%
-            }
-          }
-        }
-      }
-    }
-  }
-}
+//package com.wavesplatform.state.diffs
```
Comment (vsuharnikov) [inline node/src/test/scala/com/wavesplatform/state/diffs/BlockDifferDetailedDiffTest.scala]: Uncomment or remove?

File: node/src/main/scala/com/wavesplatform/transaction/smart/WavesEnvironment.scala (hunk 8 above)
Comment (xrtm000) [reply node/src/main/scala/com/wavesplatform/transaction/smart/WavesEnvironment.scala:358]: That's for the future)

File: node/src/main/scala/com/wavesplatform/database/RocksDBWriter.scala (hunk 1 above)
Comment (xrtm000) [reply node/src/main/scala/com/wavesplatform/database/RocksDBWriter.scala]: Done

File: node/src/test/scala/com/wavesplatform/test/DomainPresets.scala (hunk 2 above)
Comment (xrtm000) [reply node/src/test/scala/com/wavesplatform/test/DomainPresets.scala]: Done

File: node/src/main/scala/com/wavesplatform/state/StateSnapshot.scala (hunk 3 above)
Comment (xrtm000) [reply node/src/main/scala/com/wavesplatform/state/StateSnapshot.scala]: Done

File: node/src/main/scala/com/wavesplatform/state/VolumeAndFee.scala (hunk 4 above)
Comment (xrtm000) [reply node/src/main/scala/com/wavesplatform/state/VolumeAndFee.scala]: Done

File: node/src/main/scala/com/wavesplatform/database/Caches.scala (hunk 6 above)
Comment (xrtm000) [reply node/src/main/scala/com/wavesplatform/database/Caches.scala]: Done

File: node/src/test/scala/com/wavesplatform/state/diffs/smart/predef/BrokenUnicodeTest.scala (hunk 7 above)
Comment (xrtm000) [reply node/src/test/scala/com/wavesplatform/state/diffs/smart/predef/BrokenUnicodeTest.scala]: Restored

File: node/src/test/scala/com/wavesplatform/state/snapshot/TxStateSnapshotHashSpec.scala [hunk 10]
```diff
@@ -107,7 +109,8 @@ class TxStateSnapshotHashSpec extends PropSpec with WithDomain {
 
   property("correctly create transaction state snapshot hash from diff") {
```
Comment (ivan-mashonskiy) [inline node/src/test/scala/com/wavesplatform/state/snapshot/TxStateSnapshotHashSpec.scala]: * from snapshot

File: node-it/src/test/scala/com/wavesplatform/it/sync/smartcontract/EstimatorTestSuite.scala (hunk 5 above)
Comment (xrtm000) [reply node-it/src/test/scala/com/wavesplatform/it/sync/smartcontract/EstimatorTestSuite.scala:129]: Done

File: node/src/test/scala/com/wavesplatform/state/diffs/ScriptComplexityCountTest.scala [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+import com.wavesplatform.transaction.transfer.MassTransferTransaction.ParsedTransfer
+import com.wavesplatform.transaction.{GenesisTransaction, Transaction, TxNonNegativeAmount, TxVersion}
+import org.scalatest.Inside
+
+// noinspection NameBooleanParameters
+class ScriptComplexityCountTest extends PropSpec with WithDomain with Inside {
+  val script = TestCompiler(V6).compileExpression("func f() = true\n f()")
+  // complexity = 1 for both estimator and evaluator
+
+  property("check scripts run count") {
```
Comment (ivan-mashonskiy) [inline node/src/test/scala/com/wavesplatform/state/diffs/ScriptComplexityCountTest.scala]: *complexity

Comment (xrtm000) [reply node/src/test/scala/com/wavesplatform/state/diffs/ScriptComplexityCountTest.scala]: Done

**Review by ivan-mashonskiy — APPROVED** [review]

**Review by phearnot — APPROVED** [review]

**Pull request merged**